### ISSUE TITLE
Add a Flag Type Syntax

### ISSUE THREAD

**Issue by MohammadHamdyGhanem** [dotnet/csharplang#1323]
I came across many dissuasions about enhancing enums, in languages and CoreFx repos. Many of them concerns easily setting and unsetting bits in  enums marked as flags. I think adding a Flag Type maybe a solution:

#Defenetion:
```
Flag Color: int
{
    Red,
    Green,
    Blue
}
```

#Usage:
```
var myColor = new Color(false, true, true);
myColor.Red = true;
myColor.Green = false;
myColor.Blue = ! myColor.Blue;
If (myColor.Blue)
   foo();

int i = myColor;
myColor = 10;
```

#Implementation:
The compiler should transform Flag syntax into this generated class:

```
        class Color
        {
           public enum ColorEnum: int
            {
                None = 0,
                Red = 1,
                Green = 2,
                Blue = 4
            }

            ColorEnum Value; 

            public Color()
            {
                Value = ColorEnum.None;
            }

            public Color(bool red, bool green, bool blue)
            {
                Red = red;
                Green = green;
                Blue = blue;
            }

            public bool Red
            {
                get => (Value & ColorEnum.Red) == ColorEnum.Red;
                set
                {
                    if (value)
                        Value |= ColorEnum.Red;
                    else
                        Value &= ~ColorEnum.Red;
                }
            }

            public bool Green
            {
                get => (Value & ColorEnum.Green) == ColorEnum.Green;
                set
                {
                    if (value)
                        Value |= ColorEnum.Green;
                    else
                        Value &= ~ColorEnum.Green;
                }
            }

            public bool Blue
            {
                get => (Value & ColorEnum.Blue) == ColorEnum.Blue;
                set
                {
                    if (value)
                        Value |= ColorEnum.Blue;
                    else
                        Value &= ~ColorEnum.Blue;
                }
            }

            Enum ToEnum() => Value;

            public static implicit operator int (Color color) => (int) color.Value;
            public static implicit operator Color(int value) => new Color() {
                Red = ((int)ColorEnum.Red & value) > 0,
                Green = ((int)ColorEnum.Green & value) > 0,
                Blue = ((int)ColorEnum.Blue & value) > 0
            };

        }
```

#Benefits:
There is no breaking change. The Flag code is done by the generated class. If the Flag is used anyway like an enum (as in corefx method calls),  C# can convert this code to the Color.ColorEnum to talk to the CoreFx in regular way.

**Comment by jnm2** [comment]
> #Defenetion

I read that as ‘defenestration’ just now.

**Comment by ufcpp** [comment]
You can make a code generator for that like https://github.com/ufcpp/BitFields, without any new language syntax.

**Comment by MohammadHamdyGhanem** [comment]
@ufcpp 
This code generator will not be a public solution. Not all programmers understand binary system and bitwise operations. Making Flag operations easier will help them. There is a proposal in VB to add a syntax for flag operations:
https://github.com/dotnet/vblang/issues/228
and in coreFx to add methods to set flags on and off:
https://github.com/dotnet/corefx/issues/15453

**Comment by CyrusNajmabadi** [comment]
I don't understand the example at all:

```c#
var myColor = new Color(false, true, true);
myColor.Red = true;
myColor.Green = false;
myColor.Blue = ! myColor.Blue;
```

What does this mean?  I've never seen an API where a Color is defined as 3 booleans.  Seems extremely strange.  

Do you have a real-world use case where this would be helpful?

**Comment by yaakov-h** [comment]
Sounds a bit like #1303.

**Comment by DavidArno** [comment]
> Not all programmers understand binary system and bitwise operations

Not all programmers need to use bitwise operations. When they do, they should learn about binary and how bitwise operations work. Syntactic sugar or other "magic" systems should never be an excuse to not bother to learn about how things work under the hood. 

So the only measure of the worth of this proposal should whether, from the perspective of someone who already understands flag-based enums, does this proposal make them easier to work with.

So by replacing:
```cs
myColor |= Color.Red;
myColor &= ~Color.Green;
```
with:
```cs
myColor.Red = true;
myColor.Green = false;
```
have we made flag-based enums easier to work with? Maybe, but it's a tiny improvement in my view.

**Comment by DavidArno** [comment]
@yaakov-h,

Hmm, I'd say it's a direct duplicate of #1303. Nice find.

**Comment by MohammadHamdyGhanem** [comment]
I first proposed this Idea two months ago, in this comment:
https://github.com/dotnet/vblang/issues/228#issuecomment-353989126
The #1303 is opened 11 days ago, and I didn't seee it. I refined my idea after new !+ operator proposed in VB that I don't like.
The Color example is not mine, but it started in https://github.com/dotnet/vblang/issues/228# and propagated :)

**Comment by MohammadHamdyGhanem** [comment]
I add some operators to the Flag class to allow bitwise operations as well. now one can do compose a yellow flag like this:
`var Yellow = new Color (false, true, true);`
 Or
`var Yellow = new Color(){Green = true, Blue = true};`
 Or:
`Color Yellow = Color.ColorEnum.Green | Color.ColorEnum.Blue;`
 Or:
`var Yellow = new Color() {Green = true} | new Color() {Blue= true};`

**Comment by MgSam** [comment]
@DavidArno I agree no language changes are needed, but are you seriously suggested that this is readable? 

``` C#
myColor |= Color.Red;
myColor &= ~Color.Green;
if (myColor & Color.Blue) ...
```

How often do people do bit-wise operations? Almost never. This code works, obviously, but it does not express the intent in any readable manner. I much prefer using extension methods for this purpose.

**Comment by CyrusNajmabadi** [comment]
> I agree no language changes are needed, but are you seriously suggesting that this is readable?

Yes.  That seems like very idiomatic C-style bit manipulation.  I'm not sure if we could do anything better than ```|=``` for merging bits.  ```x &= ~y``` is wordy for to disable bits.  So maybe we could have something like ```x ~= y``` which was defined to do the same thing.

**Comment by CyrusNajmabadi** [comment]
> I add some operators to the Flag class to allow bitwise operations as well. now one can do compose a yellow flag like this:
> var Yellow = new Color (false, true, true);

that doesn't seem any better than:

```c#
enum Color
{
    Red, Green, Blue, Yellow = Green | Blue
}
```

**Comment by CyrusNajmabadi** [comment]
> now one can do compose a yellow flag like this:

Why is this composition desirable?  Whereas there was one way to do things before ```x | y```, you've now made 4 different ways that one has to learn that are *more* verbose, but end up boiling down to the same thing.  How is that an improvement for people?

**Comment by jnm2** [comment]
> Color Yellow = Color.ColorEnum.Green | Color.ColorEnum.Blue;

Don't you mean `yellow = green | red`, if you're doing additive mixing, or yellow = green & ~blue` if you're talking subtractive?

**Comment by MohammadHamdyGhanem** [comment]
> Why is this composition desirable?

**Comment by DavidArno** [comment]
@MgSam,

I've with @CyrusNajmabadi here. Having used C-style bitwise manipulation on and off for 30 years, I find such expressions easy to read. 😉 

But I'd love support for `~=` as I agree that `&= ~` is clunky.

**Comment by MohammadHamdyGhanem** [comment]
So, what if we overload the + and - operators in the Enum class to do the trick like this:```

        public static Enum operator +(Enum value1, object value2)
        {
            return (Enum)(object)((int)(object)value1 | (int)(object)value2);
        }

        public static Enum operator -(Enum value1, object value2) 
		{
            return (Enum)(object)((int)(object)value1 & ~(int)(object)value2);
        }
```

**Comment by jnm2** [comment]
I think that would be too confusing. There are already places where I treat enums as their numeric value.

I am partial to `|=` and `~=`, even though `~=` would be a departure from the symmetry we have so far.

**Comment by whoisj** [comment]
> So, what if we overload the + and - operators in the Enum class to do the trick like this:

Because only a small portion of `enum` values are flags.

**Comment by MohammadHamdyGhanem** [comment]
The code can be modified to check if the enum has a flag attribute. If it's a flag then do the bitwise operations, if not , use the uasual arrithmitic operations.

**Comment by jnm2** [comment]
> The code can be modified to check if the enum has a flag attribute. If it's a flag then do the bitwise operations, if not , use the uasual arrithmitic operations.

This is even more confusing.

**Comment by MohammadHamdyGhanem** [comment]
`This is even more confusing.`

**Comment by CyrusNajmabadi** [comment]
I agree.  Having + mean numeric addition in one place, and in anotehr mean 'bitwise-or'ing, is just asking for trouble.

**Comment by CyrusNajmabadi** [comment]
why do you need

```myColor += Color.Red;```

when you already have

```myColor |= Color.Red;```

?

+ already has meaning for numeric oriented constructs (which enums **definitely** are in .Net).  Giving it a very different meaning in this domain is bad idea.

**Comment by jnm2** [comment]
> So, at the basic binary level, we are familiar with the concept that Or = +

At the basic binary level, 1 + 1 does not equal 1.

**Comment by MohammadHamdyGhanem** [comment]
As said before, bitwise operations are confusing to the programmer who doesn't understand binary system. I failed once to explane to some students why must they use 
if ((e.Button & MouseButtons.Left) == MouseButtons.Left)
instead of 
if (e.Button == MouseButtons.Left)
Whithout teaching them binary system first, whuch was not possible!
This is a general purpose hi level language, and should get red of any parts that need low level knowlage! 
With my suggestion, this will be:
if (e.Button.Left)
A self explainig statement.

**Comment by jnm2** [comment]
> This is a general purpose hi level language, and should get red of any parts that need low level knowlage!

I'll let [C# 7.2](https://docs.microsoft.com/dotnet/csharp/whats-new/csharp-7-2) know. 😆

**Comment by MohammadHamdyGhanem** [comment]
Please hurry :D

**Comment by CyrusNajmabadi** [comment]
> This is a general purpose high level language, and should get red of any parts that need low level knowleage!

Sorry, that's an ok stance to take when deciding on what C# 1.0 should be.  But we're on 8.0 now.  Enums have been part and parcel of .net for 15+ years.  We're not going to get rid of them.

**Comment by CyrusNajmabadi** [comment]
> As said before, bitwise operations are confusing to the programmer who doesn't understand binary system. 

And what then happens when the programmer doesn't understand booleans?  Do we get rid of those as well?  What if they don't understand virtual dispatch?  Remove also?  What about value types?  

We are not going to remove important parts of .net just because a new developer didn't understand them.

**Comment by whoisj** [comment]
> As said before, bitwise operations are confusing to the programmer who doesn't understand binary system. 

Oh well, I'm going to go out on a limb here and say a "programmer who doesn't understand binary" is likely having a lot of issues with writing compiled code. Perhaps they ought to learn how things work. I'm not certain any amount of syntactic sugar is going to help them. In fact, too much might give them enough rope to hang themselves with - not a friendly things to do.

**Comment by AdamSpeight2008** [comment]
>The code can be modified to check if the enum has a flag attribute. 

Using the proposed overloading the + - operators in the `Enum` code, would involve a run-time reflection verification. The syntax I proposed in https://github.com/dotnet/vblang/issues/228 does this check at compile-time. Plus allows us to do constant-folding.
